File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/am_to_amap.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

// Ham chuyen tu ma tran ke sang ban do ke (adjacency map)
map<int, map<int, int>> adjacencyMatrixToMap(const vector<vector<int>>& matrix) {
    int n = matrix.size();
    map<int, map<int, int>> adjMap;  // Ban do ke

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            int weight = matrix[i][j];
            if (weight != 0) {
                adjMap[i][j] = weight;  // Gan dinh j voi trong so vao map cua i
            }
        }
    }

    return adjMap;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;

    vector<vector<int>> matrix(n, vector<int>(n));
    cout << "Nhap ma tran ke:\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> matrix[i][j];

    auto adjMap = adjacencyMatrixToMap(matrix);

    // In ket qua
    cout << "Adjacency Map:\n";
    for (const auto& [u, neighbors] : adjMap) {
        cout << u << ": ";
        for (const auto& [v, w] : neighbors) {
            cout << "(" << v << ", " << w << ") ";
        }
        cout << endl;
    }

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/am_to_amap.py
# Ham chuyen ma tran ke sang ban do ke (adjacency map)
def adjacency_matrix_to_map(matrix):
    n = len(matrix)
    adj_map = {}  # Khoi tao dict chinh

    for i in range(n):  # Duyet tung dinh i
        neighbors = {}  # Dict con chua dinh ke cua i
        for j in range(n):  # Duyet tung dinh j
            weight = matrix[i][j]
            if weight != 0:  # Neu co canh tu i den j
                neighbors[j] = weight  # Ghi lai j voi trong so
        adj_map[i] = neighbors  # Gan danh sach ke vao i

    return adj_map

# Nhap du lieu tu ban phim
n = int(input("Nhap so dinh: "))
matrix = []
print("Nhap ma tran ke:")
for _ in range(n):
    row = list(map(int, input().split()))
    matrix.append(row)

adj_map = adjacency_matrix_to_map(matrix)

# In ket qua
print("Adjacency Map:")
for u in adj_map:
    print(f"{u}: {adj_map[u]}")